### PR TITLE
Run fit

### LINKED ISSUE #4: Complete test run Sage 2.2.1 on small dataset

### PR DESCRIPTION
**Pull request by jaclark5**
Closes #4 

This PR is currently compared to #14 and should be compared to `main` before merging. This step must wait until #14 is complete.

### CODE REVIEW

**Comment by lilyminium** [comment]
I know this is still in draft status, but just FYI this is an intimidatingly large PR to review. `fit_data.py` does multiple things -- could you please split them up? (Correspondingly, could you please split #4 up as sub-issues?) There seem to be three distinct steps here that create I/O for the next step that would be natural stopping points:
- split_train_test, which creates the training/testing directories
- prepare_to_train, which creates the smeetensorforcefield and topologies. This one's not incredibly obvious, but just from experience this can be a really slow step that's convenient to save out in case something goes wrong in training, or you want to play with hyperparameters, loss, regularisation etc etc. Looking at the log file you can see it takes ~a second per Interchange -- with the complete training dataset of around 23k, that's 6+ hours. One other reason splitting this up is helpful is that parameterising can happen on a CPU queue which is charged at 1 SU/hr, whereas training needs a CUDA GPU which is charged at 34 SU/hr. I usually dump these out as a pickle because it's easy.
- actually training the FF on GPU

**Comment by jaclark5** [comment]
@lilyminium thanks for the insight!

File: 6_run_fit/test_run/fit_data.py [hunk 1]
```diff
@@ -0,0 +1,601 @@
...
+        # the include/exclude types are Interchange PotentialKey.id's -- typically SMIRKS
+        # include=[], <-- bonds to train. Not specifying trains all
+        # exclude=[], <-- bonds NOT to train
+    ),
+    #    "Angles": descent.train.ParameterConfig(
+    #        cols=["k", "angle"],
+    #        scales={"k": 1e-2, "angle": 1.0},
+    #        limits={"k": [0.0, None], "angle": [0.0, math.pi]},
+    #    ),
+}
```
Comment (Yoshanuikabundi) [inline 6_run_fit/test_run/fit_data.py:100]: Does this mean "Only train bonds" or "train bonds, angles, propers and impropers, but use the default settings for everything other than bonds"? I'm like 75% sure it's the former...

File: 6_run_fit/test_run/fit_data.py [hunk 2]
```diff
@@ -0,0 +1,601 @@
...
+            # now update the force field
+            for smirks, k_s in collection_data.items():
+                ff_parameter = handler[smirks]
+                k_mapped_to_p = [k_s[p] for p in ff_parameter.periodicity]
+                ff_parameter.k = k_mapped_to_p
+
+        elif handler_name in ["ImproperTorsions"]:
+            k_index = parameter_attrs.index("k")
+            handler = starting_ff.get_parameter_handler(handler_name)
+            # we only fit the v2 terms for improper torsions so convert to list and set
```
Comment (Yoshanuikabundi) [inline 6_run_fit/test_run/fit_data.py:446]: Should keep this behavior in mind in case we try to fit some other impropers

**Review by Yoshanuikabundi — APPROVED** [review]
This LGTM! I love your docstrings, though I will note that in Infra-land we're moving away from specifying types in docstrings so that there's no duplication of information between docstring and type annotations - docstrings tend to go out of date.

The only thing I've found that could be interpreted as incorrect/erroneous/mistaken is that the docstring for `train_forcefield` claims the loss function is a sum over squared differences, but I think it's implemented as a mean... very minor point.

Other than that, my comments are mostly just entirely optional notes that you should feel free to ignore about ways to make the code marginally more readable, plus maybe a question or two where I'm not super clear on what's going on (which is a me-problem, not a your-code-problem). I am about to write my fitting code so thanks for letting me take a look at yours!

I'm also very liable to get science things wrong/not notice scientific mistakes - my research and training has been in applying force fields, not fitting them - so please don't rely too heavily on my tick of approval for that sort of thing.
